### PR TITLE
Fully type certbot-nginx module

### PR DESCRIPTION
**Pull request by adferrand**
This PR adds types to all functions in certbot-nginx module.

### CODE REVIEW

File: certbot-nginx/certbot_nginx/_internal/parser.py
Comment (alexzorin) [inline certbot-nginx/certbot_nginx/_internal/parser.py:431]: Maybe? Or just `UnspacedList` and wrap the `[]` case?

```suggestion
def _parse_ssl_options(ssl_options: Optional[str]) -> Union[UnspacedList, List[Any]]:
```

File: certbot-nginx/certbot_nginx/_internal/parser.py
Comment (alexzorin) [inline certbot-nginx/certbot_nginx/_internal/parser.py:116]: Since it is imported, we may as well do this across the file?

```suggestion
    def _get_raw_servers(self) -> Dict[str, Union[List[Any], UnspacedList]]:
```

**Review by alexzorin — CHANGES_REQUESTED** [review]
I tried to give this a review but I'm finding it hard to navigate all the `List[Any]`/`Any`. I think it will need a second review from somebody more familiar with the code.

File: certbot-nginx/certbot_nginx/_internal/configurator.py
Comment (atombrella) [inline certbot-nginx/certbot_nginx/_internal/configurator.py:930]: `Union[str, List[str], None]`: Just asking a question, because I'm not an expert in typing, but isn't this a case for `Optional[str, List[str]]`

File: certbot-nginx/certbot_nginx/_internal/configurator.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/configurator.py:223]: I typed based on the original signature of the configurator class here, but I agree with you that the interface do not declare `chain_path` and `fullchain_path` as `Optional[str]` and maybe more importantly, they are supposed to be positional.

I avoided as much as possible to modify the actual behavior of the code to start with, because it would have made a lot harder the typing if the errors can come from the typing itself or mistake I introduce by modifying the actual behavior.

Now that it is settled, indeed we can start to talk about the discrepencies that the typing is revealing. Here it is one, the signature of the configurator do not represent the actual usages: `chain_path` and `fullchain_path` are always set. Moreover the implementation shows that it is not expected to be something else. I also checked the `install` phase, and the others, that come down to the `certbot._internal.client.install_certificate()` function that always set the `chain_path` and the `fullchain_path` to the underlying intaller plugin.

This and the fact `certbot-nginx` code is not part of the public API convince me to modify now the signatures and make them coherent with the interfaces.

File: certbot-nginx/certbot_nginx/_internal/configurator.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/configurator.py:443]: Going in that direction, the most elegant approach would be to go to a dataclass, that I find superior to embed a structured and typed set of elements. Dataclasses are a little cumbersome to define on Python <3.7, but as soon as we drop Python 3.6 we can take advantage of @dataclass decorator (https://docs.python.org/3/library/dataclasses.html). On Python 3.7, we can take advantage of `typing_extension` module that exposes `TypeDict` also.

Globally my preference would go to a dataclass here, what do you think ?

File: certbot-nginx/certbot_nginx/_internal/nginxparser.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/nginxparser.py:80]: Nothing specific from the types, but `pyparsing.core.ParserElement.parse_string` documentation gives no room to imagine anything else than a list of string. And the consumption of the type added in `as_list()` seems coherent with how it is consumed later.

File: certbot-nginx/certbot_nginx/_internal/nginxparser.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/nginxparser.py:144]: I would say that it is a good case for the `@overload`! Let me try that.

File: certbot-nginx/certbot_nginx/_internal/obj.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/obj.py:170]: Yes, callers of the `path` parameters seem to agree.

File: certbot-nginx/certbot_nginx/_internal/configurator.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/configurator.py:930]: You mean `Optional[Union[str, List[str]]]`. Well, technically a `Optional[T]` is a `Union[None, T]` in fact. Most of the time `Optional` is more elegant, but if we are talking about a `Union` type, I find a little more elegant to use only two levels of generics (here `Union` and `List`) instead of three (`Union`, `List`  and `Optional`).

For sure PEP 604 is way more elegant, because we could use `List[str] | str | None`, but it is native to Python 3.10. For the other Python versions, we need `typing_extensions` module and declare `from __future__ import annotations`. I am not that in a hurry to go back to our lovely `from __future__ import absolute_import, print_function` of the era of Python 2->3 migration ^^

Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/configurator.py:930]: Buuuut I am saying that and used `Optional[Union[str, List[str]]]` in `enhance()`... So I will be consistent and use `Optional[Union[str, List[str]]]` in `_enable_ocsp_stapling()`.

File: certbot-nginx/certbot_nginx/_internal/configurator.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/configurator.py]: Sure, fixed that!

**Review by ohemorange — CHANGES_REQUESTED** [review]
Phew. There's maybe only a couple actually necessary things here, mostly a bunch of tightening up types based on my understanding of code flow.

File: certbot-nginx/certbot_nginx/_internal/configurator.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/configurator.py:457]: In this PR, as for the previous ones concerning certbot typing, I try to stick with the robustness principle (https://en.wikipedia.org/wiki/Robustness_principle): be conservative with what you do, be liberal with what you accept. Which translates here by using the most possible generic type for the parameters of the functions, and the most exact type for the actual returned values.

Here basically, we need a container object that is iterable and sizable. This corresponds to a `Sequence`, for which `Tuple` and `List` are acceptable actual types. The type contained must be something that can access data by named indexes, which corresponds to a `Mapping`, and `Dict` is an acceptable actual type.

Another way to consider the problem is to define if the types are here to describe the intention of usage defined by the developer, or the technically possible usages permitted by the implementation. I had a related discussion on that matter with @bmw in https://github.com/certbot/josepy/pull/122, and we went on considering the technically possible usages in types.

File: certbot-nginx/certbot_nginx/_internal/configurator.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/configurator.py:477]: See my reasonning on https://github.com/certbot/certbot/pull/9124#discussion_r778450161 to keep the `Iterable`.

File: certbot-nginx/certbot_nginx/_internal/configurator.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/configurator.py:515]: See my reasonning on https://github.com/certbot/certbot/pull/9124#discussion_r778450161 to keep the `Iterable`.

File: certbot-nginx/certbot_nginx/_internal/configurator.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/configurator.py:366]: Sadly, mypy is reporting that `ipv6_info()` is still called with an `int` parameter in `certbot-nginx/certbot_nginx/_internal/http_01.py:130` and `certbot-nginx/certbot_nginx/_internal/configurator.py:717`.

File: certbot-nginx/certbot_nginx/_internal/nginxparser.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/nginxparser.py:81]: Ok, so the typing here is strongly related to the way `UnspacedList` is constructed.

To properly define that, we would need recursive types (like `NestedListStr = List[Union[str, 'NestedListStr']]`, but this is not a feature supported by mypy.

So the "best" way to type here is do use `List[Any]` as with `UnspacedList`.

File: certbot-nginx/certbot_nginx/_internal/configurator.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/configurator.py:939]: Because the signature of the function itself is associated to the way the enhance logic works.

At high level this function, and the other functions related to enhance, are grouped in the dict `_enhance_func`. The effective function to call is accessed by index and then called with the parameters provided by `enhance()`. The signature of `enhance()` itself is controlled by the `Installer` interface, that `NginxConfigurator` implements, which dictates that `options`, the argument 2 passed to enhance functions, is of type `Optional[Union[List[str], str]]`.

So add the end from mypy perspective, all enhance functions must accept `Optional[Union[List[str], str]]`, and the fact that `_enable_ocsp_stapling` raise an error if the option is a `List[str]` is an implementation detail.

File: certbot-nginx/certbot_nginx/_internal/configurator.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/configurator.py:1030]: The return of this function is generated by `tuple(int(i) for i in product_version.split("."))` which gives no clue about how many `int` we will have in the tuple at the end.

Of course we could raise an exception if the tuple does not contain 3 ints, and make mypy happy when declaring `Tuple[int, int, int]`.

However from an external perspective, type `Tuple[int, ...]` is sufficient to do a correct comparison with another `Tuple[int, ...]` of a different size, and so it is compatible (and so sufficient) for the expected usage (doing version comparison).

File: certbot-nginx/certbot_nginx/_internal/display_ops.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/display_ops.py:14]: See my reasonning on https://github.com/certbot/certbot/pull/9124#discussion_r778450161 to keep the `Iterable`.

File: certbot-nginx/certbot_nginx/_internal/http_01.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/http_01.py:121]: Always a good occasion here to also fix the doc!

File: certbot-nginx/certbot_nginx/_internal/nginxparser.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/nginxparser.py:220]: Here the type of `__setitem__` is dictated by the protocol around setters. The fact that slice is not supported and raise an error is an implementation detail.

File: certbot-nginx/certbot_nginx/_internal/nginxparser.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/nginxparser.py:229]: See above.

File: certbot-nginx/certbot_nginx/_internal/nginxparser.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/nginxparser.py]: Thanks!

File: certbot-nginx/certbot_nginx/_internal/http_01.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/http_01.py:124]: Because it turns out that the Nginx implementation of `Addr.fromstring()` can return `None`: 
```
        # Ignore UNIX-domain sockets
        if addr.startswith('unix:'):
            return None
```

So I propagate this information accordingly to callers.

File: certbot-nginx/certbot_nginx/_internal/http_01.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/http_01.py:147]: As a consequence of the above, we want to clear the `None` elements in `addresses` before we return from `_default_listen_addresses()`.

Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/http_01.py:147]: Basically here mypy uncovered a flaw: without appropriate nullity checks, `default_listen_addresses()` may return a list containing `None` elements, that could break `_make_server_block()` because we call `addr.to_string(include_default=False)` to each `addr` generated in the list returned by `default_listen_addresses()`.

File: certbot-nginx/certbot_nginx/_internal/nginxparser.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/nginxparser.py:154]: Indeed examples in PEP-484 is providing examples where the actual implementation is untyped, and there is the sentence "while the latter is used at runtime but should be ignored by a type checker" (where the "latter" refers here to the actual implementation). But I think that what PEP-484 is saying that any type in the actual implementation should be ignored for variable inference, but not that the actual implementation must be untyped itself.

Effectively mypy, with `--disallow-untyped-defs` set, is raising an error if the actual implementation is not properly typed.

File: certbot-nginx/certbot_nginx/_internal/nginxparser.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/nginxparser.py:170]: I answer here for `insert`, `append` and `extend`. These three functions have their type dictated by the superclass (`List`) which says that the actual type of `x` must be the same than the generic of the `List`. So to make x be of type `Union[str, None, List[Any]]`, I need to make `UnspacedList` heritate from `List[Union[str, None, List[Any]]]` instead of `List[Any]`.

This generates around 50 errors in the type checking basically because of the recursive type problem we already discussed in https://github.com/certbot/certbot/pull/9124#discussion_r778453975. Quite similarly, we have to stick with `List[Any]`, which consequently leads `x` to be `Any` in `insert`, `append`  and `extend`.

File: certbot-nginx/certbot_nginx/_internal/nginxparser.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/nginxparser.py:179]: See above.

File: certbot-nginx/certbot_nginx/_internal/nginxparser.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/nginxparser.py:187]: See above.

File: certbot-nginx/certbot_nginx/_internal/nginxparser.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/nginxparser.py]: Oh sorry. It was not my intention to add an actual implementation to this function. I removed that.

File: certbot-nginx/certbot_nginx/_internal/configurator.py
Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/configurator.py:443]: @alexzorin since this is still theoritical as we have to support Python 3.6 and 3.7, I consider this discussion closed, but do not hesitate to comment otherwise.

**Comment by adferrand** [comment]
@ohemorange thanks for your review. I finished answering to it!

File: certbot-nginx/certbot_nginx/_internal/configurator.py
Comment (ohemorange) [reply certbot-nginx/certbot_nginx/_internal/configurator.py:366]: that's great, mypy found a bug! therefore we should make this change, and fix the code in those places. or, if we prefer, keep it as an int -- there's no reason it should accept both types, but either should be fine.

File: certbot-nginx/certbot_nginx/_internal/configurator.py
Comment (ohemorange) [reply certbot-nginx/certbot_nginx/_internal/configurator.py:457]: Ok, that sounds good to me, as long as we're being intentional about it.

File: certbot-nginx/certbot_nginx/_internal/configurator.py
Comment (ohemorange) [reply certbot-nginx/certbot_nginx/_internal/configurator.py:939]: thanks for tracing that!

File: certbot-nginx/certbot_nginx/_internal/configurator.py
Comment (ohemorange) [reply certbot-nginx/certbot_nginx/_internal/configurator.py:366]: any thoughts on which way to handle this?

Comment (adferrand) [reply certbot-nginx/certbot_nginx/_internal/configurator.py]: Sorry I missed that. I think `str` is completely fine. I fixed the callers where `int` was used.

**Pull request merged**
